src/fairseq2/recipes/lm/preference_units/dpo_unit.py

```diff
@@ -0,0 +1,184 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Tuple, Union, cast, final
+
+import torch
+import torch.distributed
+from torch import Tensor
+from torch.nn import Module
+from torcheval.metrics import Mean
+
+from fairseq2.datasets.preference import PreferenceOptimizationBatch
+from fairseq2.gang import Gang, get_rank
+from fairseq2.logging import get_log_writer
+from fairseq2.metrics.recorder import format_as_float, register_metric_formatter
+from fairseq2.models.sequence import (
+    SequenceBatch,
+    SequenceModelOutput,
+    as_auto_regressive_input,
+)
+from fairseq2.recipes.common_metrics import SequenceMetricBag
+from fairseq2.recipes.lm.preference_units.preference_criterion_config import (
+    PreferenceCriterionConfig,
+)
+from fairseq2.recipes.trainer import AbstractTrainUnit
+from fairseq2.typing import DataType, override
+
+log = get_log_writer(__name__)
+
+
+@dataclass
+class DpoFinetuneConfig(PreferenceCriterionConfig):
+    """Holds the DPO-finetuning configuration of a language model."""
+
+    # Hyperparameters
+    dpo_beta: float = 0.1
+    """The coefficient of regularization towards the reference model."""
+
+    nll_scale: float = 0.0
+    """The coefficient of NLL loss added to the DPO loss."""
+
+    # Reference Model
+    reference_model: Union[str, Path] = "llama3_8b_instruct"
+    """The name or path to the asset card of the reference model to use."""
+
+    reference_dtype: DataType = torch.bfloat16
+    """The data type of the reference model."""
+
+    reference_tensor_parallel_size: int = 1
+    """The size of tensor parallelism for the reference model."""
+
+
+@final
+class DpoFinetuneUnit(AbstractTrainUnit[PreferenceOptimizationBatch]):
+    """Represents the DPO-finetuning unit of a language model."""
+
+    _reference_model: Module
+    _beta: float
+    _nll_scale: float
+    _metric_bag: DpoFinetuneMetricBag
+
+    def __init__(
+        self,
+        model: Module,
+        reference_model: Union[Module | None],
+        gang: Gang,
+        beta: float = 0.1,
+        nll_scale: float = 1.0,
+    ) -> None:
+        super().__init__(model)
+
+        assert (
```

Comment:
Why is `reference_model` parameter declared as optional? DPO always expects to have a reference model.